LND: add neutrino backend support

PatMulligan opened the pull request:
Add option to use neutrino instead of bitcoind for fetching blockchain data. This allows running LND as a lightweight client that connects to remote Bitcoin full nodes.

New options:
- services.lnd.backend: choose between "bitcoind" (default) or "neutrino"
- services.lnd.neutrino.addpeers: list of Bitcoin node peers to connect to
- services.lnd.neutrino.feeUrl: fee estimation URL

Example configuration:

```nix
  services.lnd = {
    enable = true;

    backend = "neutrino";
    neutrino.addpeers = [
      "btcd1.lnolymp.us"
      "btcd2.lnolymp.us"
      "btcd-mainnet.lightning.computer"
      "uswest.blixtwallet.com"
      "node.eldamar.icu"
      "noad.sathoarder.com"
      "bb1.breez.technology"
      "bb2.breez.technology"
      "neutrino.shock.network"
    ];

    neutrino.feeUrl = "https://nodes.lightning.computer/fees/v1/btc-fee-estimates.json" # default value

  };
```

PatMulligan committed:
lnd: add neutrino backend support

Add option to use neutrino instead of bitcoind for fetching blockchain data. This allows running LND as a lightweight client that connects to remote Bitcoin full nodes.

New options:
- services.lnd.backend: choose between "bitcoind" (default) or "neutrino"
- services.lnd.neutrino.addpeers: list of Bitcoin node peers to connect to
- services.lnd.neutrino.feeUrl: fee estimation URL

prusnak commented on modules/lnd.nix:
```diff
@@ -144,6 +144,30 @@ let
+      description = "The backend to use for fetching blockchain data.";
+    };
+    neutrino = {
+      addpeers = mkOption {
```
wondering whether this should rather be `peers` instead of `addpeers`?

PatMulligan replied:
yeah i think it makes sense!

prusnak commented on modules/lnd.nix:
```diff
@@ -288,7 +324,7 @@ in {
       isSystemUser = true;
       group = cfg.group;
-      extraGroups = [ "bitcoinrpc-public" ];
+      extraGroups = optional (cfg.backend == "bitcoind") "bitcoinrpc-public";
```
`optional` is very cryptic and it is not obvious it is adding an element into the list, suggesting using `optionals` instead:

```suggestion
      extraGroups = optionals (cfg.backend == "bitcoind") [ "bitcoinrpc-public" ];
```

PatMulligan committed:
use optionals instead of optional for list additions

Use optionals with explicit list syntax for requires, after, and extraGroups

PatMulligan committed:
Rename neutrino.addpeers to neutrino.peers

Use simpler naming for the NixOS option. The option defines which peers to connect to, not the action of adding them. At least 1 required.

PatMulligan committed:
Add neutrino.maxPeers option

Allow configuring the maximum number of inbound and outbound peers for neutrino. Default is 8, matching lnd's default.

PatMulligan committed:
add neutrino backend example for lnd

ekzyis commented on modules/lnd.nix:
```diff
@@ -144,6 +144,37 @@ let
+    neutrino = {
+      peers = mkOption {
+        type = types.listOf types.str;
+        default = [];
```
nit: could add defaults here, too, or was it intentionally left out? I see there's a default for `feeUrl`.

ekzyis commented on modules/lnd.nix:
```diff
@@ -144,6 +144,37 @@ let
       description = "LND TLS certificate path.";
     };
     tor = nbLib.tor;
+    backend = mkOption {
+      type = types.enum [ "bitcoind" "neutrino" ];
+      default = "bitcoind";
+      description = "The backend to use for fetching blockchain data.";
+    };
```
nit: I think `lnd.neutrino.enable` xor `lnd.bitcoind.enable` would be more common instead of a new `lnd.backend` option.

ekzyis reviewed (approved):
I'm running this [rebased on master](https://github.com/ekzyis/nix-bitcoin/tree/add-lnd-neutrino) and my new LND v0.20.1-beta node synced fine. Neutrino config written to /var/lib/lnd/lnd.conf is as expected.

Left two nits, else LGTM!

> should I add any more config options or maybe just let it be handled in services.lnd.extraConfig...?

I think the existing basic config is fine. No need to attempt to include all config options. I really like how simple this was to setup. Thank you for this PR! You saved me time.
